### PR TITLE
Support field projection via REST API

### PR DESCRIPTION
**Pull request by kostko**
Adds support for projections like this:
* `data` (top-level projection)
* `name` (top-level projection)
* `descriptor_schema__name` (nested projection)
* `data__name` (nested projection)
* `descriptor_schema__schema__name` (nested projection + projection into JSON field)
* `data__input__reads` (nested projection + projection into JSON field)
* `data__output__bam__file` (nested projection + deep projection into JSON field)

TODO:
* [x] Tests
* [x] Documentation

### CODE REVIEW

**Comment by kostko** [comment]
Added tests, this is ready for review.

File: resolwe/flow/serializers.py [hunk 1]
```diff
@@ -145,24 +157,26 @@ class Meta:
...
-        super(DataSerializer, self).__init__(*args, **kwargs)
-
-        request = kwargs.get('context', {}).get('request', None)
+    def get_fields(self):
+        """Dynamically adapt fields based on the current request."""
+        fields = super(DataSerializer, self).get_fields()
 
-        if not hasattr(request, 'method') or request.method == "GET":
-            self.fields['descriptor_schema'] = DescriptorSchemaSerializer(required=False)
+        if not hasattr(self.request, 'method') or self.request.method == "GET":
```
Comment (dblenkus) [inline resolwe/flow/serializers.py]: Is `not hasattr(self.request, 'method')` still required? This was a workaround while observers didn't pass the context.

Same applies for other serializers.

File: resolwe/rest/projection.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+FIELD_SEPARATOR = ','
+FIELD_DEREFERENCE = '__'
+
+
+def apply_subfield_projection(field, value, deep=False):
+    """Apply projection from request context.
+
+    The passed dictionary may be mutated.
+
+    :param field: An instance of `Field` or `Serializer`
```
Comment (dblenkus) [inline resolwe/rest/projection.py]: Add type to params.

**Comment by kostko** [comment]
I've now also added documentation for field projections.

**Pull request merged**